# Computer Vision Nanodegree

## Project: Image Captioning

---

In this notebook, you will train your CNN-RNN model.  

You are welcome and encouraged to try out many different architectures and hyperparameters when searching for a good model.

This does have the potential to make the project quite messy!  Before submitting your project, make sure that you clean up:
- the code you write in this notebook.  The notebook should describe how to train a single CNN-RNN architecture, corresponding to your final choice of hyperparameters.  You should structure the notebook so that the reviewer can replicate your results by running the code in this notebook.  
- the output of the code cell in **Step 2**.  The output should show the output obtained when training the model from scratch.

This notebook **will be graded**.  

Feel free to use the links below to navigate the notebook:
- [Step 1](#step1): Training Setup
- [Step 2](#step2): Train your Model
- [Step 3](#step3): (Optional) Validate your Model

<a id='step1'></a>
## Step 1: Training Setup

In this step of the notebook, you will customize the training of your CNN-RNN model by specifying hyperparameters and setting other options that are important to the training procedure.  The values you set now will be used when training your model in **Step 2** below.

You should only amend blocks of code that are preceded by a `TODO` statement.  **Any code blocks that are not preceded by a `TODO` statement should not be modified**.

### Task #1

Begin by setting the following variables:
- `batch_size` - the batch size of each training batch.  It is the number of image-caption pairs used to amend the model weights in each training step. 
- `vocab_threshold` - the minimum word count threshold.  Note that a larger threshold will result in a smaller vocabulary, whereas a smaller threshold will include rarer words and result in a larger vocabulary.  
- `vocab_from_file` - a Boolean that decides whether to load the vocabulary from file. 
- `embed_size` - the dimensionality of the image and word embeddings.  
- `hidden_size` - the number of features in the hidden state of the RNN decoder.  
- `num_epochs` - the number of epochs to train the model.  We recommend that you set `num_epochs=3`, but feel free to increase or decrease this number as you wish.  [This paper](https://arxiv.org/pdf/1502.03044.pdf) trained a captioning model on a single state-of-the-art GPU for 3 days, but you'll soon see that you can get reasonable results in a matter of a few hours!  (_But of course, if you want your model to compete with current research, you will have to train for much longer._)
- `save_every` - determines how often to save the model weights.  We recommend that you set `save_every=1`, to save the model weights after each epoch.  This way, after the `i`th epoch, the encoder and decoder weights will be saved in the `models/` folder as `encoder-i.pkl` and `decoder-i.pkl`, respectively.
- `print_every` - determines how often to print the batch loss to the Jupyter notebook while training.  Note that you **will not** observe a monotonic decrease in the loss function while training - this is perfectly fine and completely expected!  You are encouraged to keep this at its default value of `100` to avoid clogging the notebook, but feel free to change it.
- `log_file` - the name of the text file containing - for every step - how the loss and perplexity evolved during training.

If you're not sure where to begin to set some of the values above, you can peruse [this paper](https://arxiv.org/pdf/1502.03044.pdf) and [this paper](https://arxiv.org/pdf/1411.4555.pdf) for useful guidance!  **To avoid spending too long on this notebook**, you are encouraged to consult these suggested research papers to obtain a strong initial guess for which hyperparameters are likely to work best.  Then, train a single model, and proceed to the next notebook (**3_Inference.ipynb**).  If you are unhappy with your performance, you can return to this notebook to tweak the hyperparameters (and/or the architecture in **model.py**) and re-train your model.

Some of the 

### Question 1

**Question:** Describe your CNN-RNN architecture in detail.  With this architecture in mind, how did you select the values of the variables in Task 1?  If you consulted a research paper detailing a successful implementation of an image captioning model, please provide the reference.

**Answer:** 

I consulted the two papers listed above "Show and Tell: A Neural Image Caption Generator" and "Show, Attend and Tell: Neural Image Caption Generation with Visual Attention" to gain a better understanding on the overall process and framework to implement an Image Caption.

The overall approach includes an Encoder and a Decoder. The **Encoder** leveraged a pre-trained CNN architecture like `ResNet-50` to help extract features from the images. The line `modules = list(resnet.children())[:-1]` is when we remove the last Fully Connected layer in the pre-trained architecture. The goal of the encoder is to enode the content of the image into a smaller feature vector. Which later gets passed to the Decoder (RNN) network.

The **Decoder** which is our LSTM that we use to generate the captions for an image. We pass the encoder embedding as input to the LSTM to learn. The vocabulary in the trainnig pool are pretty much all the unique words in our data set, with the added <start> and <end> tokens to indicate start and end of a sentence. 
    
I used a `vecob_threshold` of 4. The number of words were better than 5, and I did not want to increase the vocab words to include every rare word that may have only been used once. A balance would be key, so I went with 4. If I would test and rerun the training again, I would like to try 2-3 to evaluate the generated captions and the quality of the output.

for the `batch_size` I think 10 was too small. If I had the bandwitdh to rerun this again, I would probably increase this value and the size of each batch to maybe 32.

`embed_size` is used both in embedded image feature vector and word embedding. In this case I used 256 and a `hidden_size` of 512.

### (Optional) Task #2

Note that we have provided a recommended image transform `transform_train` for pre-processing the training images, but you are welcome (and encouraged!) to modify it as you wish.  When modifying this transform, keep in mind that:
- the images in the dataset have varying heights and widths, and 
- if using a pre-trained model, you must perform the corresponding appropriate normalization.

### Question 2

**Question:** How did you select the transform in `transform_train`?  If you left the transform at its provided value, why do you think that it is a good choice for your CNN architecture?

**Answer:** 
The `transform_train` was not changed. It covered resizing, random cropping, random horizontal flip, converting to tensor, followed by normalization. This is a standard image transformer that I felt was more than sufficient for the current task. These transformations are what is needed for the `ResNet50` since it is a pre-trained model we need to make sure the images are in the expected size and normal form.

### Task #3

Next, you will specify a Python list containing the learnable parameters of the model.  For instance, if you decide to make all weights in the decoder trainable, but only want to train the weights in the embedding layer of the encoder, then you should set `params` to something like:
```
params = list(decoder.parameters()) + list(encoder.embed.parameters()) 
```

### Question 3

**Question:** How did you select the trainable parameters of your architecture?  Why do you think this is a good choice?

**Answer:** 
For the trainable parameters I made all the weights int he decoder trainable, which at every iteration they get updated. For the encoder, I used to only updated the weights in the emedding layer. 

### Task #4

Finally, you will select an [optimizer](http://pytorch.org/docs/master/optim.html#torch.optim.Optimizer).

### Question 4

**Question:** How did you select the optimizer used to train your model?

**Answer:** 
I selected Adam optimizer, instead of SGD. From past experiences I have seen it perform/converge faster. When I first started the training I used the wrong learning rate of 0.01 which was too high and there was no improvement and had to end it after 1 epoch. I updated the Learning Rate to 0.001 and noticed signifcant improvement in the overall learning.

<a id='paso1'></a>.
## Paso 1: Configuración del entrenamiento

En este paso del libro de trabajo, personalizarás el entrenamiento de tu modelo CNN-RNN especificando hiperparámetros y configurando otras opciones importantes para el procedimiento de entrenamiento.  Los valores que establezcas ahora se utilizarán cuando entrenes tu modelo en el **Paso 2**.

Sólo debe modificar los bloques de código que están precedidos por una declaración `ALL`.  **Cualquier bloque de código que no esté precedido por una sentencia «ALL» no debe ser modificado**.

### Tarea #1

Comience estableciendo las siguientes variables:
- `batch_size` - el tamaño de cada lote de entrenamiento.  Este es el número de pares imagen-capítulo utilizados para modificar los pesos del modelo en cada paso de entrenamiento. 
- `vocab_threshold` - el umbral mínimo de recuento de palabras.  Tenga en cuenta que un umbral más alto resultará en un vocabulario más pequeño, mientras que un umbral más bajo incluirá palabras más raras y resultará en un vocabulario más grande.  
- `vocab_from_file` - un booleano que decide si se carga el vocabulario desde un archivo. 
- `embed_size` - la dimensionalidad de la imagen y de las palabras incrustadas.  
- `hidden_size` - el número de características en el estado oculto del decodificador RNN.  
- `num_epochs` - número de épocas para entrenar el modelo.  Le recomendamos que establezca `num_epochs=3`, pero siéntase libre de aumentar o disminuir este número como desee.  [Este artículo](https://arxiv.org/pdf/1502.03044.pdf) entrenó un modelo de subtitulación en una única GPU de última generación durante 3 días, ¡pero pronto verás que puedes obtener resultados razonables en cuestión de unas pocas horas! (_Pero, por supuesto, si quieres que tu modelo compita con la investigación actual, tendrás que entrenar durante mucho más tiempo._)
- `save_every` - determina la frecuencia con la que se guardan los pesos del modelo.  Te recomendamos que configures `save_every=1`, para guardar los pesos del modelo después de cada epoch.  De esta forma, después de la época `i`, los pesos del codificador y decodificador se guardarán en la carpeta `models/` como `encoder-i.pkl` y `decoder-i.pkl`, respectivamente.
- `print_every` - determina la frecuencia con la que se imprimirá la pérdida por lotes en el cuaderno Jupyter durante el entrenamiento.  Tenga en cuenta que **no** observará una disminución monótona en la función de pérdida durante el entrenamiento - ¡esto está perfectamente bien y es de esperar!  Se recomienda mantener este valor en su valor por defecto de «100» para evitar la obstrucción del cuaderno, pero no dude en cambiarlo.
- `log_file` - el nombre del archivo de texto que contiene - para cada paso - cómo la pérdida y la perplejidad evolucionaron durante el entrenamiento.

Si no estás seguro de por dónde empezar a configurar algunos de los valores anteriores, puedes consultar [este documento](https://arxiv.org/pdf/1502.03044.pdf) y [este documento](https://arxiv.org/pdf/1411.4555.pdf) para obtener una guía útil.  **Para evitar dedicar demasiado tiempo a este cuaderno**, le animamos a consultar estos documentos de investigación sugeridos para obtener una estimación inicial sólida sobre qué hiperparámetros es probable que funcionen mejor.  A continuación, entrene un único modelo y pase al siguiente cuaderno (**3_Inference.ipynb**).  Si no está satisfecho con su rendimiento, puede volver a este cuaderno para ajustar los hiperparámetros (y/o la arquitectura en **model.py**) y volver a entrenar su modelo.

Algunas de las 

### Pregunta 1

**Pregunta:** Describe detalladamente tu arquitectura CNN-RNN.  Dada esta arquitectura, ¿cómo seleccionaste los valores de las variables en la Tarea 1?  Si has consultado un trabajo de investigación que detalle una implementación exitosa de un modelo de subtitulado de imágenes, proporciona la referencia.

**Respuesta. 

Consulté los dos artículos mencionados anteriormente «Show and Tell: A Neural Image Caption Generator» y «Show, Attend and Tell: Neural Image Caption Generation with Visual Attention» para comprender mejor el proceso general y el marco para implementar el subtitulado de imágenes.

El enfoque general incluye un codificador y un decodificador. El **codificador** utiliza una arquitectura CNN preentrenada como `ResNet-50` para ayudar a extraer características de las imágenes. La línea `modules = list(resnet.children())[:-1]` es donde eliminamos la última capa totalmente conectada de la arquitectura pre-entrante. El objetivo del codificador es codificar el contenido de la imagen en un vector de características más pequeño. Que más tarde se pasa a la Red Decodificador (RNN).

El **Decodificador** es nuestro LSTM que utilizamos para generar subtítulos para una imagen. Pasamos la incrustación del codificador como entrada al LSTM para que aprenda. El vocabulario en la piscina trainnig es más o menos todas las palabras únicas en nuestro conjunto de datos, con el añadido <start> y <end> tok.

Traducción realizada con la versión gratuita del traductor DeepL.com

In [4]:
import torch
import torch.nn as nn
from torchvision import transforms
import sys
sys.path.append('./opt/cocoapi/PythonAPI')
sys.path.append('C:/ws/scm/github/mia/image-captioning-pytorch/opt/cocoapi/PythonAPI')
from pycocotools.coco import COCO
from data_loader import get_loader
from model import EncoderCNN, DecoderRNN
import math


## Select appropriate values for the Python variables below.
batch_size = 10          # batch size
vocab_threshold = 4        # minimum word count threshold
vocab_from_file = True      # if True, load existing vocab file
embed_size = 256           # dimensionality of image and word embeddings
hidden_size = 512          # number of features in hidden state of the RNN decoder
num_epochs = 3             # number of training epochs
save_every = 1             # determines frequency of saving model weights
print_every = 100          # determines window for printing average loss
log_file = 'training_log.txt'       # name of file with saved training loss and perplexity

# (Optional)  Amend the image transform below.
transform_train = transforms.Compose([ 
    transforms.Resize(256),                          # smaller edge of image resized to 256
    transforms.RandomCrop(224),                      # get 224x224 crop from random location
    transforms.RandomHorizontalFlip(),               # horizontally flip image with probability=0.5
    transforms.ToTensor(),                           # convert the PIL Image to a tensor
    transforms.Normalize((0.485, 0.456, 0.406),      # normalize image for pre-trained model
                         (0.229, 0.224, 0.225))])

# Build data loader.
data_loader = get_loader(transform=transform_train,
                         mode='train',
                         batch_size=batch_size,
                         vocab_threshold=vocab_threshold,
                         vocab_from_file=vocab_from_file)

# The size of the vocabulary.
vocab_size = len(data_loader.dataset.vocab)

# Initialize the encoder and decoder. 
encoder = EncoderCNN(embed_size)
decoder = DecoderRNN(embed_size, hidden_size, vocab_size)

# Move models to GPU if CUDA is available. 
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
encoder.to(device)
decoder.to(device)

# Define the loss function. 
criterion = nn.CrossEntropyLoss().cuda() if torch.cuda.is_available() else nn.CrossEntropyLoss()

# Specify the learnable parameters of the model.
params = list(decoder.parameters())+list(encoder.embed.parameters())

# Define the optimizer.
optimizer = torch.optim.Adam(params,lr=0.001)

# Set the total number of training steps per epoch.
total_step = math.ceil(len(data_loader.dataset.caption_lengths) / data_loader.batch_sampler.batch_size)

Vocabulary successfully loaded from vocab.pkl file!
loading annotations into memory...
Done (t=0.59s)
creating index...
index created!
Obtaining caption lengths...


100%|██████████| 414113/414113 [00:16<00:00, 25340.96it/s]
c:\ws\scm\github\mia\MIA203-NN-DL\.venv\Lib\site-packages\torchvision\models\_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and may be removed in the future, please use 'weights' instead.
  warnings.warn(
c:\ws\scm\github\mia\MIA203-NN-DL\.venv\Lib\site-packages\torchvision\models\_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and may be removed in the future. The current behavior is equivalent to passing `weights=ResNet50_Weights.IMAGENET1K_V1`. You can also use `weights=ResNet50_Weights.DEFAULT` to get the most up-to-date weights.
  warnings.warn(msg)


<a id='step2'></a>
## Step 2: Train your Model

Once you have executed the code cell in **Step 1**, the training procedure below should run without issue.  

It is completely fine to leave the code cell below as-is without modifications to train your model.  However, if you would like to modify the code used to train the model below, you must ensure that your changes are easily parsed by your reviewer.  In other words, make sure to provide appropriate comments to describe how your code works!  

You may find it useful to load saved weights to resume training.  In that case, note the names of the files containing the encoder and decoder weights that you'd like to load (`encoder_file` and `decoder_file`).  Then you can load the weights by using the lines below:

```python
# Load pre-trained weights before resuming training.
encoder.load_state_dict(torch.load(os.path.join('./models', encoder_file)))
decoder.load_state_dict(torch.load(os.path.join('./models', decoder_file)))
```

While trying out parameters, make sure to take extensive notes and record the settings that you used in your various training runs.  In particular, you don't want to encounter a situation where you've trained a model for several hours but can't remember what settings you used :).

### A Note on Tuning Hyperparameters

To figure out how well your model is doing, you can look at how the training loss and perplexity evolve during training - and for the purposes of this project, you are encouraged to amend the hyperparameters based on this information.  

However, this will not tell you if your model is overfitting to the training data, and, unfortunately, overfitting is a problem that is commonly encountered when training image captioning models.  

For this project, you need not worry about overfitting. **This project does not have strict requirements regarding the performance of your model**, and you just need to demonstrate that your model has learned **_something_** when you generate captions on the test data.  For now, we strongly encourage you to train your model for the suggested 3 epochs without worrying about performance; then, you should immediately transition to the next notebook in the sequence (**3_Inference.ipynb**) to see how your model performs on the test data.  If your model needs to be changed, you can come back to this notebook, amend hyperparameters (if necessary), and re-train the model.

That said, if you would like to go above and beyond in this project, you can read about some approaches to minimizing overfitting in section 4.3.1 of [this paper](http://ieeexplore.ieee.org/stamp/stamp.jsp?arnumber=7505636).  In the next (optional) step of this notebook, we provide some guidance for assessing the performance on the validation dataset.

In [6]:
import torch.utils.data as data
import numpy as np
import os
import requests
import time

# Open the training log file.
f = open(log_file, 'w')

old_time = time.time()
# response = requests.request("GET", 
#                             "http://metadata.google.internal/computeMetadata/v1/instance/attributes/keep_alive_token", 
#                             headers={"Metadata-Flavor":"Google"})

for epoch in range(1, num_epochs+1):
    
    for i_step in range(1, total_step+1):
        
        # if time.time() - old_time > 60:
        #     old_time = time.time()
        #     requests.request("POST", 
        #                      "https://nebula.udacity.com/api/v1/remote/keep-alive", 
        #                      headers={'Authorization': "STAR " + response.text})
        
        # Randomly sample a caption length, and sample indices with that length.
        indices = data_loader.dataset.get_train_indices()
        # Create and assign a batch sampler to retrieve a batch with the sampled indices.
        new_sampler = data.sampler.SubsetRandomSampler(indices=indices)
        data_loader.batch_sampler.sampler = new_sampler
        
        # Obtain the batch.
        images, captions = next(iter(data_loader))

        # Move batch of images and captions to GPU if CUDA is available.
        images = images.to(device)
        captions = captions.to(device)
        
        # Zero the gradients.
        decoder.zero_grad()
        encoder.zero_grad()
        
        # Pass the inputs through the CNN-RNN model.
        features = encoder(images)
        outputs = decoder(features, captions)
        
        # Calculate the batch loss.
        loss = criterion(outputs.view(-1, vocab_size), captions.view(-1))
        
        # Backward pass.
        loss.backward()
        
        # Update the parameters in the optimizer.
        optimizer.step()
            
        # Get training statistics.
        stats = 'Epoch [%d/%d], Step [%d/%d], Loss: %.4f, Perplexity: %5.4f' % (epoch, num_epochs, i_step, total_step, loss.item(), np.exp(loss.item()))
        
        # Print training statistics (on same line).
        print('\r' + stats, end="")
        sys.stdout.flush()
        
        # Print training statistics to file.
        f.write(stats + '\n')
        f.flush()
        
        # Print training statistics (on different line).
        if i_step % print_every == 0:
            print('\r' + stats)
            
    # Save the weights.
    if epoch % save_every == 0:
        torch.save(decoder.state_dict(), os.path.join('./models', 'decoder-%d.pkl' % epoch))
        torch.save(encoder.state_dict(), os.path.join('./models', 'encoder-%d.pkl' % epoch))

# Close the training log file.f.close()

Epoch [1/3], Step [100/41412], Loss: 4.3546, Perplexity: 77.8339
Epoch [1/3], Step [200/41412], Loss: 3.6881, Perplexity: 39.96827
Epoch [1/3], Step [300/41412], Loss: 3.4995, Perplexity: 33.09816
Epoch [1/3], Step [400/41412], Loss: 4.1127, Perplexity: 61.11254
Epoch [1/3], Step [500/41412], Loss: 3.4724, Perplexity: 32.21566
Epoch [1/3], Step [600/41412], Loss: 3.4137, Perplexity: 30.3768
Epoch [1/3], Step [700/41412], Loss: 3.5903, Perplexity: 36.24348
Epoch [1/3], Step [800/41412], Loss: 3.4253, Perplexity: 30.7329
Epoch [1/3], Step [900/41412], Loss: 3.1002, Perplexity: 22.2035
Epoch [1/3], Step [1000/41412], Loss: 2.9904, Perplexity: 19.8946
Epoch [1/3], Step [1100/41412], Loss: 3.4756, Perplexity: 32.3162
Epoch [1/3], Step [1200/41412], Loss: 3.1154, Perplexity: 22.5415
Epoch [1/3], Step [1300/41412], Loss: 3.2224, Perplexity: 25.0870
Epoch [1/3], Step [1400/41412], Loss: 2.8149, Perplexity: 16.6914
Epoch [1/3], Step [1500/41412], Loss: 3.1198, Perplexity: 22.6421
Epoch [1/3], S

<a id='step3'></a>
## Step 3: (Optional) Validate your Model

To assess potential overfitting, one approach is to assess performance on a validation set.  If you decide to do this **optional** task, you are required to first complete all of the steps in the next notebook in the sequence (**3_Inference.ipynb**); as part of that notebook, you will write and test code (specifically, the `sample` method in the `DecoderRNN` class) that uses your RNN decoder to generate captions.  That code will prove incredibly useful here. 

If you decide to validate your model, please do not edit the data loader in **data_loader.py**.  Instead, create a new file named **data_loader_val.py** containing the code for obtaining the data loader for the validation data.  You can access:
- the validation images at filepath `'/opt/cocoapi/images/train2014/'`, and
- the validation image caption annotation file at filepath `'/opt/cocoapi/annotations/captions_val2014.json'`.

The suggested approach to validating your model involves creating a json file such as [this one](https://github.com/cocodataset/cocoapi/blob/master/results/captions_val2014_fakecap_results.json) containing your model's predicted captions for the validation images.  Then, you can write your own script or use one that you [find online](https://github.com/tylin/coco-caption) to calculate the BLEU score of your model.  You can read more about the BLEU score, along with other evaluation metrics (such as TEOR and Cider) in section 4.1 of [this paper](https://arxiv.org/pdf/1411.4555.pdf).  For more information about how to use the annotation file, check out the [website](http://cocodataset.org/#download) for the COCO dataset.

<a id='paso3'></a>
## Paso 3: (Opcional) Validar el modelo
Para evaluar el sobreajuste potencial, un enfoque es evaluar el rendimiento en un conjunto de validación.  Si decide realizar esta tarea **opcional**, primero deberá completar todos los pasos en el siguiente cuaderno de la secuencia (**3_Inference.ipynb**); como parte de ese cuaderno, escribirá y probará código (específicamente, el método `sample` en la clase `DecoderRNN`) que utiliza su decodificador RNN para generar subtítulos.  Ese código será increíblemente útil aquí. 
Si decides validar tu modelo, por favor no edites el cargador de datos en **data_loader.py**.  En su lugar, crea un nuevo archivo llamado **data_loader_val.py** que contenga el código para obtener el cargador de datos para los datos de validación.  Puedes acceder a
- las imágenes de validación en la ruta de archivo `'/opt/cocoapi/images/train2014/'`, y
- el archivo de anotación de la imagen de validación en la ruta de archivo `'/opt/cocoapi/annotations/captions_val2014.json'`.
El método sugerido para validar el modelo consiste en crear un archivo json como [éste](https://github.com/cocodataset/cocoapi/blob/master/results/captions_val2014_fakecap_results.json) que contenga los títulos previstos del modelo para las imágenes de validación.  A continuación, puedes escribir tu propio script o utilizar uno que [encuentres en línea](https://github.com/tylin/coco-caption) para calcular la puntuación BLEU de tu modelo.  Puede leer más sobre la puntuación BLEU, junto con otras métricas de evaluación (como TEOR y Cider) en la sección 4.1 de [este documento](https://arxiv.org/pdf/1411.4555.pdf).  Para obtener más información sobre cómo utilizar el archivo de anotación, consulte el [sitio web](http://cocodataset.org/#download) del conjunto de datos COCO.


Traducción realizada con la versión gratuita del traductor DeepL.com

<a id='paso3'></a>
## Paso 3: (Opcional) Validar su modelo

Para evaluar el sobreajuste potencial, un enfoque es evaluar el rendimiento en un conjunto de validación.  Si decide realizar esta tarea **opcional**, primero deberá completar todos los pasos en el siguiente cuaderno de la secuencia (**3_Inference.ipynb**); como parte de ese cuaderno, escribirá y probará código (específicamente, el método `sample` en la clase `DecoderRNN`) que utiliza su decodificador RNN para generar subtítulos.  Ese código será increíblemente útil aquí. 

Si decides validar tu modelo, por favor no edites el cargador de datos en **data_loader.py**.  En su lugar, crea un nuevo archivo llamado **data_loader_val.py** que contenga el código para obtener el cargador de datos para los datos de validación.  Puedes acceder a
- las imágenes de validación en la ruta de archivo `'/opt/cocoapi/images/train2014/'`, y
- el archivo de anotación de la imagen de validación en la ruta de archivo `'/opt/cocoapi/annotations/captions_val2014.json'`.

El método sugerido para validar el modelo consiste en crear un archivo json como [éste](https://github.com/cocodataset/cocoapi/blob/master/results/captions_val2014_fakecap_results.json) que contenga los títulos previstos del modelo para las imágenes de validación.  A continuación, puedes escribir tu propio script o utilizar uno que encuentres en Internet (https://github.com/tylin/coco-caption) para calcular la puntuación BLEU de tu modelo.  Puede leer más sobre la puntuación BLEU, junto con otras métricas de evaluación (como TEOR y Cider) en la sección 4.1 de [este documento](https://arxiv.org/pdf/1411.4555.pdf).  Para obtener más información sobre cómo utilizar el archivo de anotación, consulte el [sitio web](http://cocodataset.org/#download) del conjunto de datos COCO.